doc/man/man1/rtrace.1
.\" RCSid "$Id$"
.TH RTRACE 1 10/17/97 RADIANCE
.SH NAME
rtrace - trace rays in RADIANCE scene
.SH SYNOPSIS
.B rtrace
[
.B options
]
[
.B $EVAR
]
[
.B @file
]
.B octree
.br
.B "rtrace [ options ] \-defaults"
.SH DESCRIPTION
.I Rtrace
traces rays from the standard input through the RADIANCE scene given by
.I octree
and sends the results to the standard output.
(The octree may be given as the output of a command enclosed in quotes
and preceded by a `!'.)\0
Input for each ray is:

	xorg yorg zorg xdir ydir zdir

If the direction vector is (0,0,0), a bogus record
is printed and the output is flushed if the
.I -x
value is one or zero.
(See the notes on this option below.)\0
This may be useful for programs that run
.I rtrace
as a separate process.
In the second form, the default values
for the options (modified by those options present)
are printed with a brief explanation.
.PP
Options may be given on the command line and/or read from the
environment and/or read from a file.
A command argument beginning with a dollar sign ('$') is immediately
replaced by the contents of the given environment variable.
A command argument beginning with an at sign ('@') is immediately
replaced by the contents of the given file.
Most options are followed by one or more arguments, which must be
separated from the option and each other by white space.
The exceptions to this rule are the boolean options.
Normally, the appearance of a boolean option causes a feature to
be "toggled", that is switched from off to on or on to off
depending on its previous state.
Boolean options may also be set
explicitly by following them immediately with a '+' or '-', meaning
on or off, respectively.
Synonyms for '+' are any of the characters "yYtT1", and synonyms
for '-' are any of the characters "nNfF0".
All other characters will generate an error.
.TP 10n
.BI -f io
Format input according to the character
.I i
and output according to the character
.I o.
.I Rtrace
understands the following input and output formats:  'a' for
ascii, 'f' for single-precision floating point,
and 'd' for double-precision floating point.
In addition to these three choices, the character 'c' may be used
to denote 4-byte floating point (Radiance) color format
for the output of values only
.I (\-ov
option, below).
If the output character is missing, the input format is used.
.IP
Note that there is no space between this option and its argument.
.TP
.BI -o spec
Produce output fields according to
.I spec.
Characters are interpreted as follows:
.IP
o	origin (input)
.IP
d	direction (normalized)
.IP
v	value (radiance)
.IP
V	contribution (radiance)
.IP
w	weight
.IP
W	color coefficient
.IP
l	effective length of ray
.IP
L	first intersection distance
.IP
c	local (u,v) coordinates
.IP
p	point of intersection
.IP
n	normal at intersection (perturbed)
.IP
N	normal at intersection (unperturbed)
.IP
s	surface name
.IP
m	modifier name
.IP
M	material name
.IP
~	tilde (end of trace marker)
.IP
If the letter 't' appears in
.I spec,
then the fields following will be printed for every ray traced,
not just the final result.
If the capital letter 'T' is given instead of 't', then all rays will
be reported, including shadow testing rays to light sources.
Spawned rays are indented one tab for each level.
The tilde marker ('~') is a handy way of differentiating the final ray
value from daughter values in a traced ray tree, and usually appears
right before the 't' or 'T' output flags.
E.g.,
.I \-ov~TmW
will emit a tilde followed by a tab at the end of each trace,
which can be easily distinguished even in binary output.
.IP
Note that there is no space between this option and its argument.
.TP
.BI -te \ mod
Append
.I mod
to the trace exclude list,
so that it will not be reported by the trace option
.I (\-o*t*).
Any ray striking an object having
.I mod
as its modifier will not be reported to the standard output with
the rest of the rays being traced.
This option has no effect unless either the 't' or 'T'
option has been given as part of the output specifier.
Any number of excluded modifiers may be given, but each
must appear in a separate option.
.TP
.BI -ti \ mod
Add
.I mod
to the trace include list,
so that it will be reported by the trace option.
The program can use either an include list or an exclude
list, but not both.
.TP
.BI -tE \ file
Same as
.I \-te,
except read modifiers to be excluded from
.I file.
The RAYPATH environment variable determines which directories are
searched for this file.
The modifier names are separated by white space in the file.
.TP
.BI -tI \ file
Same as
.I \-ti,
except read modifiers to be included from
.I file.
.TP
.BR \-i
Boolean switch to compute irradiance rather than radiance values.
This only affects the final result, substituting a Lambertian
surface and multiplying the radiance by pi.
Glass and other transparent surfaces are ignored during this stage.
Light sources still appear with their original radiance values,
though the
.I \-dv
option (below) may be used to override this.
This option is especially useful in
conjunction with ximage(1) for computing illuminance at scene points.
.TP
.BR \-u
Boolean switch to control uncorrelated random sampling.
When "off", a low-discrepancy sequence is used, which reduces
variance but can result in a brushed appearance in specular highlights.
When "on", pure Monte Carlo sampling is used in all calculations.
.TP
.BR \-I
Boolean switch to compute irradiance rather than radiance,
with the input origin and direction interpreted instead
as measurement point and orientation.
.TP
.BR \-h
Boolean switch for information header on output.
.TP
.BI -x \ res
Set the x resolution to
.I res.
The output will be flushed after every
.I res
input rays if
.I \-y
is set to zero.
A value of one means that every ray will be flushed, whatever
the setting of
.I \-y.
A value of zero means that no output flushing will take place.
.TP
.BI -y \ res
Set the y resolution to
.I res.
The program will exit after
.I res
scanlines have been processed, where a scanline is the number of rays
given by the
.I \-x
option, or 1 if
.I \-x
is zero.
A value of zero means the program will not halt until the end
of file is reached.
.IP
If both
.I \-x
and
.I \-y
options are given, a resolution string is printed at the beginning
of the output.
This is mostly useful for recovering image dimensions with
.I pvalue(1),
and for creating valid Radiance picture files using the color output
format.
(See the
.I \-f
option, above.)
.TP
.BI -n \ nproc
Execute in parallel on
.I nproc
local processes.
This option is incompatible with the
.I \-P
and
.I \-PP,
options.
Multiple processes also do not work properly with ray tree output
using any of the
.I \-o*t*
options.
There is no benefit from specifying more processes than there are
cores available on the system or the
.I \-x
setting, which forces a wait at each flush.
.TP
.BI -dj \ frac
Set the direct jittering to
.I frac.
A value of zero samples each source at specific sample points
(see the
.I \-ds
option below), giving a smoother but somewhat less accurate
rendering.
A positive value causes rays to be distributed over each
source sample according to its size, resulting in more accurate
penumbras.
This option should never be greater than 1, and may even
cause problems (such as speckle) when the value is smaller.
A warning about aiming failure will issued if
.I frac
is too large.
.TP
.BI -ds \ frac
Set the direct sampling ratio to
.I frac.
A light source will be subdivided until
the width of each sample area divided by the distance
to the illuminated point is below this ratio.
This assures accuracy in regions close to large area sources
at a slight computational expense.
A value of zero turns source subdivision off, sending at most one
shadow ray to each light source.
.TP
.BI -dt \ frac
Set the direct threshold to
.I frac.
Shadow testing will stop when the potential contribution of at least
the next and at most all remaining light sources is less than
this fraction of the accumulated value.
(See the
.I \-dc
option below.)
The remaining light source contributions are approximated
statistically.
A value of zero means that all light sources will be tested for shadow.
.TP
.BI \-dc \ frac
Set the direct certainty to
.I frac.
A value of one guarantees that the absolute accuracy of the direct calculation
will be equal to or better than that given in the
.I \-dt
specification.
A value of zero only insures that all shadow lines resulting in a contrast
change greater than the
.I \-dt
specification will be calculated.
.TP
.BI -dr \ N
Set the number of relays for secondary sources to
.I N.
A value of 0 means that secondary sources will be ignored.
A value of 1 means that sources will be made into first generation
secondary sources; a value of 2 means that first generation
secondary sources will also be made into second generation secondary
sources, and so on.
.TP
.BI -dp \ D
Set the secondary source presampling density to D.
This is the number of samples per steradian 
that will be used to determine ahead of time whether or not
it is worth following shadow rays through all the reflections and/or
transmissions associated with a secondary source path.
A value of 0 means that the full secondary source path will always
be tested for shadows if it is tested at all.
.TP
.BR \-dv
Boolean switch for light source visibility.
With this switch off, sources will be black when viewed directly
although they will still participate in the direct calculation.
This option is mostly for the program
.I mkillum(1)
to avoid inappropriate counting of light sources, but it
may also be desirable in conjunction with the
.I \-i
option.
.TP
.BI -ss \ samp
Set the specular sampling to
.I samp.
For values less than 1, this is the degree to which the highlights
are sampled for rough specular materials.
A value greater than one causes multiple ray samples to be sent
to reduce noise at a commmesurate cost.
A value of zero means that no jittering will take place, and all
reflections will appear sharp even when they should be diffuse.
.TP
.BI -st \ frac
Set the specular sampling threshold to
.I frac.
This is the minimum fraction of reflection or transmission, under which 
no specular sampling is performed.
A value of zero means that highlights will always be sampled by
tracing reflected or transmitted rays.
A value of one means that specular sampling is never used.
Highlights from light sources will always be correct, but
reflections from other surfaces will be approximated using an
ambient value.
A sampling threshold between zero and one offers a compromise between image
accuracy and rendering time.
.TP
.BR -bv
Boolean switch for back face visibility.
With this switch off, back faces of all objects will be invisible
to view rays.
This is dangerous unless the model was constructed such that
all surface normals face outward.
Although turning off back face visibility does not save much
computation time under most circumstances, it may be useful as a
tool for scene debugging, or for seeing through one-sided walls from
the outside.
.TP
.BI -av " red grn blu"
Set the ambient value to a radiance of
.I "red grn blu".
This is the final value used in place of an
indirect light calculation.
If the number of ambient bounces is one or greater and the ambient
value weight is non-zero (see
.I -aw
and
.I -ab
below), this value may be modified by the computed indirect values
to improve overall accuracy.
.TP
.BI -aw \ N
Set the relative weight of the ambient value given with the
.I -av
option to
.I N.
As new indirect irradiances are computed, they will modify the
default ambient value in a moving average, with the specified weight
assigned to the initial value given on the command and all other
weights set to 1.
If a value of 0 is given with this option, then the initial ambient
value is never modified.
This is the safest value for scenes with large differences in
indirect contributions, such as when both indoor and outdoor
(daylight) areas are visible.
.TP
.BI -ab \ N
Set the number of ambient bounces to
.I N.
This is the maximum number of diffuse bounces computed by the indirect
calculation. A value of zero implies no indirect calculation. 
.IP
This value defaults to 1 in photon mapping mode (see
.I -ap
below), implying that global photon irradiance is always computed via
.I one
ambient bounce; this behaviour applies to any positive number of ambient
bounces, regardless of the actual value specified.  A negative value enables
a preview mode that directly visualises the irradiance from the global
photon map without any ambient bounces.
.TP
.BI -ar \ res
Set the ambient resolution to
.I res.
This number will determine the maximum density of ambient values
used in interpolation.
Error will start to increase on surfaces spaced closer than
the scene size divided by the ambient resolution.
The maximum ambient value density is the scene size times the
ambient accuracy (see the
.I \-aa
option below) divided by the ambient resolution.
The scene size can be determined using
.I getinfo(1)
with the
.I \-d
option on the input octree.
.TP
.BI -aa \ acc
Set the ambient accuracy to
.I acc.
This value will approximately equal the error
from indirect illuminance interpolation.
A value of zero implies no interpolation.
.TP
.BI -ad \ N
Set the number of ambient divisions to
.I N.
The error in the Monte Carlo calculation of indirect
illuminance will be inversely proportional to the square
root of this number.
A value of zero implies no indirect calculation.
.TP
.BI -as \ N
Set the number of ambient super-samples to
.I N.
Super-samples are applied only to the ambient divisions which
show a significant change.
.TP
.BI -af \ fname
Set the ambient file to
.I fname.
This is where indirect illuminance will be stored and retrieved.
Normally, indirect illuminance values are kept in memory and
lost when the program finishes or dies.
By using a file, different invocations can share illuminance
values, saving time in the computation.
The ambient file is in a machine-independent binary format
which can be examined with
.I lookamb(1).
.IP
The ambient file may also be used as a means of communication and
data sharing between simultaneously executing processes.
The same file may be used by multiple processes, possibly running on
different machines and accessing the file via the network (ie.
.I nfs(4)).
The network lock manager
.I lockd(8)
is used to insure that this information is used consistently.
.IP
If any calculation parameters are changed or the scene
is modified, the old ambient file should be removed so that
the calculation can start over from scratch.
For convenience, the original ambient parameters are listed in the
header of the ambient file.
.I Getinfo(1)
may be used to print out this information.
.TP
.BI -ae \ mod
Append
.I mod
to the ambient exclude list,
so that it will not be considered during the indirect calculation.
This is a hack for speeding the indirect computation by
ignoring certain objects.
Any object having
.I mod
as its modifier will get the default ambient
level rather than a calculated value.
Any number of excluded modifiers may be given, but each
must appear in a separate option.
.TP
.BI -ai \ mod
Add
.I mod
to the ambient include list,
so that it will be considered during the indirect calculation.
The program can use either an include list or an exclude
list, but not both.
.TP
.BI -aE \ file
Same as
.I \-ae,
except read modifiers to be excluded from
.I file.
The RAYPATH environment variable determines which directories are
searched for this file.
The modifier names are separated by white space in the file.
.TP
.BI -aI \ file
Same as
.I \-ai,
except read modifiers to be included from
.I file.
.TP
.BI -ap " file [bwidth1 [bwidth2]]"
Enable photon mapping mode. Loads a photon map generated with
.I mkpmap(1)
from
.I file,
and evaluates the indirect irradiance depending on the photon type 
(automagically detected) using density estimates with a bandwidth of
.I bwidth1
photons, or the default bandwidth if none is specified (a warning will be
issued in this case).
.IP
Global photon irradiance is evaluated as part of the ambient calculation (see
.I \-ab
above), caustic photon irradiance is evaluated at primary rays, and 
indirect inscattering in 
.I mist
is accounted for by volume photons.
.IP
Additionally specifying 
.I bwidth2
enables bias compensation for the density estimates with a
minimum and maximum bandwidth of
.I bwidth1
and
.I bwidth2,
respectively.
.IP
Global photon irradiance may be optionally precomputed by
.I mkpmap(1),
in which case the bandwidth, if specified, is ignored, as the nearest photon
is invariably looked up.
.IP
Using direct photons replaces the direct calculation with density estimates
for debugging and validation of photon emission.       
.TP
.BI -am " frac"
Maximum search radius for photon map lookups.  Without this option, an
initial maximum search radius is estimated for each photon map from the
average photon distance to the distribution's centre of gravity.  It is then
adapted to the photon density in subsequent lookups.  This option imposes a
global fixed maximum search radius for
.I all
photon maps, thus defeating the automatic adaptation.  It is useful when
multiple warnings about short photon lookups are issued.  Note that this
option does not conflict with the bandwidth specified with the
.I \-ap
option; the number of photons found will not exceed the latter, but may be
lower if the maximum search radius contains fewer photons, thus resulting in
short lookups.  Setting this radius too large, on the other hand, may
degrade performance.
.TP
.BI -ac " pagesize"
Set the photon cache page size when using out-of-core photon mapping. The
photon cache reduces disk I/O incurred by on-demand loading (paging) of
photons, and thus increases performance. This
is expressed as a (float) multiple of the density estimate bandwidth
specified with
.I \-ap
under the assumption that photon lookups are local to a cache page. Cache
performance is sensitive to this parameter: larger pagesizes will reduce the
paging frequency at the expense of higher latency when paging does occur.
Sensible values are in the range 4 (default) to 16.
.TP
.BI -aC " cachesize"
Set the total number of photons cached when using out-of-core photon
mapping, taking into account the pagesize specified by
.I \-ac. 
Note that this is approximate as the number of cache pages is rounded to
the nearest prime. This allows adapting the cache to the available physical
memory. In conjunction with the
.I \-n
option, this is the cache size
.I per parallel process. 
Cache performance is less sensitive to this parameter,
and reasonable performance can obtained with as few as 10k photons. The
default is 1M. This option recognises multiplier suffixes (k = 1e3, M =
1e6), both in upper and lower case.
.TP
.BI -me " rext gext bext"
Set the global medium extinction coefficient to the indicated color,
in units of 1/distance (distance in world coordinates).
Light will be scattered or absorbed over distance according to
this value.
The ratio of scattering to total scattering plus absorption is set
by the albedo parameter, described below.
.TP
.BI -ma " ralb galb balb"
Set the global medium albedo to the given value between 0\00\00
and 1\01\01.
A zero value means that all light not transmitted by the medium
is absorbed.
A unitary value means that all light not transmitted by the medium
is scattered in some new direction.
The isotropy of scattering is determined by the Heyney-Greenstein
parameter, described below.
.TP
.BI \-mg \ gecc
Set the medium Heyney-Greenstein eccentricity parameter to
.I gecc.
This parameter determines how strongly scattering favors the forward
direction.
A value of 0 indicates perfectly isotropic scattering.
As this parameter approaches 1, scattering tends to prefer the
forward direction.
.TP
.BI \-ms \ sampdist
Set the medium sampling distance to
.I sampdist,
in world coordinate units.
During source scattering, this will be the average distance between
adjacent samples.
A value of 0 means that only one sample will be taken per light
source within a given scattering volume.
.TP
.BI -lr \ N
Limit reflections to a maximum of
.I N,
if N is a positive integer.
If
.I N
is zero or negative, then Russian roulette is used for ray
termination, and the
.I -lw
setting (below) must be positive.
If N is a negative integer, then this sets the upper limit
of reflections past which Russian roulette will be used.
In scenes with dielectrics and total internal reflection,
a setting of 0 (no limit) may cause a stack overflow.
.TP
.BI -lw \ frac
Limit the weight of each ray to a minimum of
.I frac.
During ray-tracing, a record is kept of the estimated contribution
(weight) a ray would have in the image.
If this weight is less than the specified minimum and the
.I -lr
setting (above) is positive, the ray is not traced.
Otherwise, Russian roulette is used to

src/rt/rtrace.c
#ifndef lint
static const char	RCSid[] = "$Id$";
#endif
/*
 *  rtrace.c - program and variables for individual ray tracing.
 */

#include "copyright.h"

/*
 *  Input is in the form:
 *
 *	xorg	yorg	zorg	xdir	ydir	zdir
 *
 *  The direction need not be normalized.  Output is flexible.
 *  If the direction vector is (0,0,0), then the output is flushed.
 *  All values default to ascii representation of real
 *  numbers.  Binary representations can be selected
 *  with '-ff' for float or '-fd' for double.  By default,
 *  radiance is computed.  The '-i' or '-I' options indicate that
 *  irradiance values are desired.
 */

#include  <time.h>

#include  "platform.h"
#include  "ray.h"
#include  "ambient.h"
#include  "source.h"
#include  "otypes.h"
#include  "resolu.h"
#include  "random.h"

extern int  inform;			/* input format */
extern int  outform;			/* output format */
extern char  *outvals;			/* output values */

extern int  imm_irrad;			/* compute immediate irradiance? */
extern int  lim_dist;			/* limit distance? */

extern char  *tralist[];		/* list of modifers to trace (or no) */
extern int  traincl;			/* include == 1, exclude == 0 */

extern int  hresolu;			/* horizontal resolution */
extern int  vresolu;			/* vertical resolution */

static int  castonly = 0;

#ifndef  MAXTSET
#define	 MAXTSET	8191		/* maximum number in trace set */
#endif
OBJECT	traset[MAXTSET+1]={0};		/* trace include/exclude set */

static RAY  thisray;			/* for our convenience */

typedef void putf_t(RREAL *v, int n);
static putf_t puta, putd, putf;

typedef void oputf_t(RAY *r);
static oputf_t  oputo, oputd, oputv, oputV, oputl, oputL, oputc, oputp,
		oputn, oputN, oputs, oputw, oputW, oputm, oputM, oputtilde;

static void setoutput(char *vs);
extern void tranotify(OBJECT obj);
static void bogusray(void);
static void raycast(RAY *r);
static void rayirrad(RAY *r);
static void rtcompute(FVECT org, FVECT dir, double dmax);
static int printvals(RAY *r);
static int getvec(FVECT vec, int fmt, FILE *fp);
static void tabin(RAY *r);
static void ourtrace(RAY *r);

static oputf_t *ray_out[16], *every_out[16];
static putf_t *putreal;


void
quit(			/* quit program */
	int  code
)
{
	if (ray_pnprocs > 0)	/* close children if any */
		ray_pclose(0);		
#ifndef  NON_POSIX
	else if (!ray_pnprocs) {
		headclean();	/* delete header file */
		pfclean();	/* clean up persist files */
	}
#endif
	exit(code);
}


char *
formstr(				/* return format identifier */
	int  f
)
{
	switch (f) {
	case 'a': return("ascii");
	case 'f': return("float");
	case 'd': return("double");
	case 'c': return(COLRFMT);
	}
	return("unknown");
}


extern void
rtrace(				/* trace rays from file */
	char  *fname,
	int  nproc
)
{
	unsigned long  vcount = (hresolu > 1) ? (unsigned long)hresolu*vresolu
					      : (unsigned long)vresolu;
	long  nextflush = (vresolu > 0) & (hresolu > 1) ? 0 : hresolu;
	FILE  *fp;
	double	d;
	FVECT  orig, direc;
					/* set up input */
	if (fname == NULL)
		fp = stdin;
	else if ((fp = fopen(fname, "r")) == NULL) {
		sprintf(errmsg, "cannot open input file \"%s\"", fname);
		error(SYSTEM, errmsg);
	}
	if (inform != 'a')
		SET_FILE_BINARY(fp);
					/* set up output */
	setoutput(outvals);
	if (imm_irrad)
		castonly = 0;
	else if (castonly)
		nproc = 1;		/* don't bother multiprocessing */
	switch (outform) {
	case 'a': putreal = puta; break;
	case 'f': putreal = putf; break;
	case 'd': putreal = putd; break;
	case 'c': 
		if (strcmp(outvals, "v"))
			error(USER, "color format with value output only");
		break;
	default:
		error(CONSISTENCY, "botched output format");
	}
	if (nproc > 1) {		/* start multiprocessing */
		ray_popen(nproc);
		ray_fifo_out = printvals;
	}
	if (hresolu > 0) {
		if (vresolu > 0)
			fprtresolu(hresolu, vresolu, stdout);
		fflush(stdout);
	}
					/* process file */
	while (getvec(orig, inform, fp) == 0 &&
			getvec(direc, inform, fp) == 0) {

		d = normalize(direc);
		if (d == 0.0) {				/* zero ==> flush */
			if (--nextflush <= 0 || !vcount) {
				if (nproc > 1 && ray_fifo_flush() < 0)
					error(USER, "lost children");
				bogusray();
				fflush(stdout);
				nextflush = (vresolu > 0) & (hresolu > 1) ? 0 :
								hresolu;
			} else
				bogusray();
		} else {				/* compute and print */
			rtcompute(orig, direc, lim_dist ? d : 0.0);
							/* flush if time */
			if (!--nextflush) {
				if (nproc > 1 && ray_fifo_flush() < 0)
					error(USER, "lost children");
				fflush(stdout);
				nextflush = hresolu;
			}
		}
		if (ferror(stdout))
			error(SYSTEM, "write error");
		if (vcount && !--vcount)		/* check for end */
			break;
	}
	if (nproc > 1) {				/* clean up children */
		if (ray_fifo_flush() < 0)
			error(USER, "unable to complete processing");
		ray_pclose(0);
	}
	if (fflush(stdout) < 0)
		error(SYSTEM, "write error");
	if (vcount)
		error(USER, "unexpected EOF on input");
	if (fname != NULL)
		fclose(fp);
}


static void
trace_sources(void)			/* trace rays to light sources, also */
{
	int	sn;
	
	for (sn = 0; sn < nsources; sn++)
		source[sn].sflags |= SFOLLOW;
}


static void
setoutput(				/* set up output tables */
	char  *vs
)
{
	oputf_t **table = ray_out;

	castonly = 1;
	while (*vs)
		switch (*vs++) {
		case 'T':				/* trace sources */
			if (!*vs) break;
			trace_sources();
			/* fall through */
		case 't':				/* trace */
			if (!*vs) break;
			*table = NULL;
			table = every_out;
			trace = ourtrace;
			castonly = 0;
			break;
		case 'o':				/* origin */
			*table++ = oputo;
			break;
		case 'd':				/* direction */
			*table++ = oputd;
			break;
		case 'v':				/* value */
			*table++ = oputv;
			castonly = 0;
			break;
		case 'V':				/* contribution */
			*table++ = oputV;
			if (ambounce > 0 && (ambacc > FTINY || ambssamp > 0))
				error(WARNING,
					"-otV accuracy depends on -aa 0 -as 0");
			break;
		case 'l':				/* effective distance */
			*table++ = oputl;
			castonly = 0;
			break;
		case 'c':				/* local coordinates */
			*table++ = oputc;
			break;
		case 'L':				/* single ray length */
			*table++ = oputL;
			break;
		case 'p':				/* point */
			*table++ = oputp;
			break;
		case 'n':				/* perturbed normal */
			*table++ = oputn;
			castonly = 0;
			break;
		case 'N':				/* unperturbed normal */
			*table++ = oputN;
			break;
		case 's':				/* surface */
			*table++ = oputs;
			break;
		case 'w':				/* weight */
			*table++ = oputw;
			break;
		case 'W':				/* coefficient */
			*table++ = oputW;
			if (ambounce > 0 && (ambacc > FTINY || ambssamp > 0))
				error(WARNING,
					"-otW accuracy depends on -aa 0 -as 0");
			break;
		case 'm':				/* modifier */
			*table++ = oputm;
			break;
		case 'M':				/* material */
			*table++ = oputM;
			break;
		case '~':				/* tilde */
			*table++ = oputtilde;
			break;
		}
	*table = NULL;
}


static void
bogusray(void)			/* print out empty record */
{
	thisray.rorg[0] = thisray.rorg[1] = thisray.rorg[2] =
	thisray.rdir[0] = thisray.rdir[1] = thisray.rdir[2] = 0.0;
	thisray.rmax = 0.0;
	rayorigin(&thisray, PRIMARY, NULL, NULL);
	printvals(&thisray);
}


static void
raycast(			/* compute first ray intersection only */
	RAY *r
)
{
	if (!localhit(r, &thescene)) {
		if (r->ro == &Aftplane) {	/* clipped */
			r->ro = NULL;
			r->rot = FHUGE;
		} else
			sourcehit(r);
	}
}


static void
rayirrad(			/* compute irradiance rather than radiance */
	RAY *r
)
{
	void	(*old_revf)(RAY *) = r->revf;

	r->rot = 1e-5;			/* pretend we hit surface */
	VSUM(r->rop, r->rorg, r->rdir, r->rot);
	r->ron[0] = -r->rdir[0];
	r->ron[1] = -r->rdir[1];
	r->ron[2] = -r->rdir[2];
	r->rod = 1.0;
					/* compute result */
	r->revf = raytrace;
	(*ofun[Lamb.otype].funp)(&Lamb, r);
	r->revf = old_revf;
}


static void
rtcompute(			/* compute and print ray value(s) */
	FVECT  org,
	FVECT  dir,
	double	dmax
)
{
					/* set up ray */
	rayorigin(&thisray, PRIMARY, NULL, NULL);
	if (imm_irrad) {
		VSUM(thisray.rorg, org, dir, 1.1e-4);
		thisray.rdir[0] = -dir[0];
		thisray.rdir[1] = -dir[1];
		thisray.rdir[2] = -dir[2];
		thisray.rmax = 0.0;
		thisray.revf = rayirrad;
	} else {
		VCOPY(thisray.rorg, org);
		VCOPY(thisray.rdir, dir);
		thisray.rmax = dmax;
		if (castonly)
			thisray.revf = raycast;
	}
	if (ray_pnprocs > 1) {		/* multiprocessing FIFO? */
		if (ray_fifo_in(&thisray) < 0)
			error(USER, "lost children");
		return;
	}
	samplendx++;			/* else do it ourselves */
	rayvalue(&thisray);
	printvals(&thisray);
}


static int
printvals(			/* print requested ray values */
	RAY  *r
)
{
	oputf_t **tp;

	if (ray_out[0] == NULL)
		return(0);
	for (tp = ray_out; *tp != NULL; tp++)
		(**tp)(r);
	if (outform == 'a')
		putchar('\n');
	return(1);
}


static int
getvec(		/* get a vector from fp */
	FVECT  vec,
	int  fmt,
	FILE  *fp
)
{
	static float  vf[3];
	static double  vd[3];
	char  buf[32];
	int  i;

	switch (fmt) {
	case 'a':					/* ascii */
		for (i = 0; i < 3; i++) {
			if (fgetword(buf, sizeof(buf), fp) == NULL ||
					!isflt(buf))
				return(-1);
			vec[i] = atof(buf);
		}
		break;
	case 'f':					/* binary float */
		if (fread((char *)vf, sizeof(float), 3, fp) != 3)
			return(-1);
		VCOPY(vec, vf);
		break;
	case 'd':					/* binary double */
		if (fread((char *)vd, sizeof(double), 3, fp) != 3)
			return(-1);
		VCOPY(vec, vd);
		break;
	default:
		error(CONSISTENCY, "botched input format");
	}
	return(0);
}


void
tranotify(			/* record new modifier */
	OBJECT	obj
)
{
	static int  hitlimit = 0;
	OBJREC	 *o = objptr(obj);
	char  **tralp;

	if (obj == OVOID) {		/* starting over */
		traset[0] = 0;
		hitlimit = 0;
		return;
	}
	if (hitlimit || !ismodifier(o->otype))
		return;
	for (tralp = tralist; *tralp != NULL; tralp++)
		if (!strcmp(o->oname, *tralp)) {
			if (traset[0] >= MAXTSET) {
				error(WARNING, "too many modifiers in trace list");
				hitlimit++;
				return;		/* should this be fatal? */
			}
			insertelem(traset, obj);
			return;
		}
}


static void
ourtrace(				/* print ray values */
	RAY  *r
)
{
	oputf_t **tp;

	if (every_out[0] == NULL)
		return;
	if (r->ro == NULL) {
		if (traincl == 1)
			return;
	} else if (traincl != -1 && traincl != inset(traset, r->ro->omod))
		return;
	tabin(r);
	for (tp = every_out; *tp != NULL; tp++)
		(**tp)(r);
	if (outform == 'a')
		putchar('\n');
}


static void
tabin(				/* tab in appropriate amount */
	RAY  *r
)
{
	const RAY  *rp;

	for (rp = r->parent; rp != NULL; rp = rp->parent)
		putchar('\t');
}


static void
oputo(				/* print origin */
	RAY  *r
)
{
	(*putreal)(r->rorg, 3);
}


static void
oputd(				/* print direction */
	RAY  *r
)
{
	(*putreal)(r->rdir, 3);
}


static void
oputv(				/* print value */
	RAY  *r
)
{
	RREAL	cval[3];

	if (outform == 'c') {
		COLR  cout;
		setcolr(cout,	colval(r->rcol,RED),
				colval(r->rcol,GRN),
				colval(r->rcol,BLU));
		fwrite(cout, sizeof(cout), 1, stdout);
		return;
	}
	cval[0] = colval(r->rcol,RED);
	cval[1] = colval(r->rcol,GRN);
	cval[2] = colval(r->rcol,BLU);
	(*putreal)(cval, 3);
}


static void
oputV(				/* print value contribution */
	RAY *r
)
{
	RREAL	contr[3];

	raycontrib(contr, r, PRIMARY);
	multcolor(contr, r->rcol);
	(*putreal)(contr, 3);
}


static void
oputl(				/* print effective distance */
	RAY  *r
)
{
	(*putreal)(&r->rt, 1);
}


static void
oputL(				/* print single ray length */
	RAY  *r
)
{
	(*putreal)(&r->rot, 1);
}


static void
oputc(				/* print local coordinates */
	RAY  *r
)
{
	(*putreal)(r->uv, 2);
}


static RREAL	vdummy[3] = {0.0, 0.0, 0.0};


static void
oputp(				/* print point */
	RAY  *r
)
{
	if (r->rot < FHUGE)
		(*putreal)(r->rop, 3);
	else
		(*putreal)(vdummy, 3);
}


static void
oputN(				/* print unperturbed normal */
	RAY  *r
)
{
	if (r->rot < FHUGE)
		(*putreal)(r->ron, 3);
	else
		(*putreal)(vdummy, 3);
}


static void
oputn(				/* print perturbed normal */
	RAY  *r
)
{
	FVECT  pnorm;

	if (r->rot >= FHUGE) {
		(*putreal)(vdummy, 3);
		return;
	}
	raynormal(pnorm, r);
	(*putreal)(pnorm, 3);
}


static void
oputs(				/* print name */
	RAY  *r
)
{
	if (r->ro != NULL)
		fputs(r->ro->oname, stdout);
	else
		putchar('*');
	putchar('\t');
}


static void
oputw(				/* print weight */
	RAY  *r
)
{
	RREAL	rwt = r->rweight;
	
	(*putreal)(&rwt, 1);
}


static void
oputW(				/* print coefficient */
	RAY  *r
)
{
	RREAL	contr[3];
				/* shadow ray not on source? */
	if (r->rsrc >= 0 && source[r->rsrc].so != r->ro)
		setcolor(contr, 0.0, 0.0, 0.0);
	else
		raycontrib(contr, r, PRIMARY);

	(*putreal)(contr, 3);
}


static void
oputm(				/* print modifier */
	RAY  *r
)
{
	if (r->ro != NULL)
		if (r->ro->omod != OVOID)
			fputs(objptr(r->ro->omod)->oname, stdout);
		else
			fputs(VOIDID, stdout);
	else
		putchar('*');
	putchar('\t');
}


static void
oputM(				/* print material */
	RAY  *r
)
{
	OBJREC	*mat;

	if (r->ro != NULL) {
		if ((mat = findmaterial(r->ro)) != NULL)
			fputs(mat->oname, stdout);
		else
			fputs(VOIDID, stdout);
	} else
		putchar('*');
	putchar('\t');
}


static void
oputtilde(			/* output tilde (spacer) */
	RAY  *r
)
{
	fputs("~\t", stdout);
}


static void
puta(				/* print ascii value(s) */
	RREAL *v, int n
)
{
	if (n == 3) {
		printf("%e\t%e\t%e\t", v[0], v[1], v[2]);
		return;
	}
	while (n--)
		printf("%e\t", *v++);
}


static void
putd(RREAL *v, int n)		/* print binary double(s) */
{
#ifdef	SMLFLT
	double	da[3];
	int	i;

	if (n > 3)
		error(INTERNAL, "code error in putd()");
	for (i = n; i--; )
		da[i] = v[i];
	fwrite(da, sizeof(double), n, stdout);
#else
	fwrite(v, sizeof(RREAL), n, stdout);
#endif
}


static void
putf(RREAL *v, int n)		/* print binary float(s) */
{
#ifndef	SMLFLT
	float	fa[3];
	int	i;

	if (n > 3)
		error(INTERNAL, "code error in putf()");
	for (i = n; i--; )
		fa[i] = v[i];
	fwrite(fa, sizeof(float), n, stdout);
#else
	fwrite(v, sizeof(RREAL), n, stdout);
#endif
}
